File: app/src/main/java/com/bachors/iptv/ChannelsActivity.java
package com.bachors.iptv;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

import com.bachors.iptv.adapter.ChannelsAdapter;
import com.bachors.iptv.models.ChannelsData;
import com.bachors.iptv.models.ChannelsResponse;
import com.bachors.iptv.util.RecyclerTouchListener;
import com.bachors.iptv.util.SharedPrefManager;
import com.google.gson.Gson;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ChannelsActivity extends AppCompatActivity {

    ProgressDialog loading;

    SharedPrefManager sharedPrefManager;
    int key;

    Context mcon;

    List<ChannelsData> allData = new ArrayList<>();
    List<ChannelsData> searchData;
    boolean all = true;
    ChannelsAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_channels);

        Bundle b = getIntent().getExtras();
        getSupportActionBar().setSubtitle(Html.fromHtml("<small>"+b.getString("title")+"</small>"));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mcon = this;

        sharedPrefManager = new SharedPrefManager(this);

        adapter = new ChannelsAdapter(this);

        RecyclerView rv = (RecyclerView) findViewById(R.id.rv);
        GridLayoutManager gl = new GridLayoutManager(this, 2);
        rv.setLayoutManager(gl);
        rv.setItemAnimator(new DefaultItemAnimator());
        rv.setAdapter(adapter);
        rv.addOnItemTouchListener(new RecyclerTouchListener(this, rv, new RecyclerTouchListener.ClickListener() {
            @Override
            public void onClick(View view, int position) {
                goTo(position);
            }

            @Override
            public void onLongClick(View view, int position) {

            }
        }));
        adapter.clear();
        jsonTogson();

    }

    private void jsonTogson() {
        try {
            JSONArray ar = new JSONArray(sharedPrefManager.getSpChannels());
            JSONObject ob = new JSONObject();
            ob.put("data", ar);
            Gson gson = new Gson();
            ChannelsResponse channels = gson.fromJson(ob.toString(), ChannelsResponse.class);
            allData = channels.getData();
            adapter.addAll(allData);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    private void filter(String text) {
        //new array list that will hold the filtered data
        searchData = new ArrayList<>();

        //looping through existing elements
        for (ChannelsData s : allData) {
            //if the existing elements contains the search input
            if (s.getName().toLowerCase().contains(text.toLowerCase())) {
                //adding the element to filtered list
                searchData.add(s);
            }
        }

        //calling a method of the adapter class and passing the filtered list
        adapter.addAll(searchData);
    }

    private void goTo(int id){
        if(all){
            key = id;
        }else{
            key = allData.indexOf(searchData.get(id));
        }

        Intent intent = new Intent(mcon, PlayerActivity.class);
        intent.putExtra("name", allData.get(key).getName());
        intent.putExtra("url", allData.get(key).getUrl());
        startActivity(intent);

    }

    @Override
    public void onBackPressed() {
        finish();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        getMenuInflater().inflate(R.menu.menu_search, menu);
        getMenuInflater().inflate(R.menu.menu_load, menu);

        MenuItem mSearch = menu.findItem(R.id.action_search);

        SearchView mSearchView = (SearchView) mSearch.getActionView();
        mSearchView.setQueryHint("Search...");

        mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                adapter.clear();
                //after the change calling the method and passing the search input
                if(s.equals("")){
                    all = true;
                    adapter.addAll(allData);
                }else {
                    all = false;
                    filter(s);
                }
                return true;
            }
        });

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem actionViewItem2 = menu.findItem(R.id.load);
        // Retrieve the action-view from menu
        View v = MenuItemCompat.getActionView(actionViewItem2);
        Button x = (Button) v.findViewById(R.id.btn_load);
        // Handle button click here
        x.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loading = ProgressDialog.show(mcon, null, "Update data ...", true, false);
                loadMain();
            }
        });
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                int id = item.getItemId();

                //noinspection SimplifiableIfStatement
                if (id == R.id.action_about) {
                    LayoutInflater inflater = this.getLayoutInflater();
                    View dialogView = inflater.inflate(R.layout.bachors_apps, null);
                    new AlertDialog.Builder(this)
                            .setCancelable(false)
                            .setNegativeButton("Ok", null)
                            .setView(dialogView)
                            .show();
                }else if (id == R.id.action_rate) {
                    Uri uri = Uri.parse("https://github.com/bachors/IPTV-Android");
                    Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                    startActivity(intent);
                }else if (id == R.id.action_share) {
                    Intent shareIntent = new Intent(Intent.ACTION_SEND);
                    shareIntent.setType("text/plain");
                    shareIntent.putExtra(Intent.EXTRA_TEXT, "https://github.com/bachors/IPTV-Android");
                    startActivity(Intent.createChooser(shareIntent, "Share link using"));
                }

                return super.onOptionsItemSelected(item);
        }
    }

    private void loadMain() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                final JSONArray json = new JSONArray();

                try {
                    Document doc = Jsoup.connect("https://raw.githubusercontent.com/iptv-org/iptv/master/README.md").get();

                    Elements tables = doc.select("table");

                    for (Element table : tables) {
                        JSONArray ar = new JSONArray();
                        Elements trs = table.select("tr");
                        int i = 0;
                        for (Element tr : trs) {
                            if(i > 0) {
                                JSONObject ob = new JSONObject();
                                ob.put("title", tr.select("td").eq(0).text());
                                ob.put("link", tr.select("td").eq(2).select("code").eq(0).text());
                                ar.put(ob);
                            }
                            i++;
                        }
                        json.put( ar);
                    }
                } catch (IOException ignored) {

                } catch (JSONException e) {
                    e.printStackTrace();
                }

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        loading.dismiss();

                        sharedPrefManager.saveSPString(SharedPrefManager.SP_PLAYLIST, json.toString());

                        Intent launchNextActivity;
                        launchNextActivity = new Intent(getApplicationContext(), ChannelsActivity.class);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                        startActivity(launchNextActivity);
                        overridePendingTransition(0, 0);
                    }
                });
            }
        }).start();
    }

}

File: app/src/main/java/com/bachors/iptv/ChannelsActivity.kt
package com.bachors.iptv

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.graphics.drawable.toDrawable
import androidx.core.view.WindowInsetsControllerCompat
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bachors.iptv.adapters.ChannelsAdapter
import com.bachors.iptv.models.ChannelsData
import com.bachors.iptv.utils.SharedPrefManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import androidx.core.net.toUri

class ChannelsActivity : AppCompatActivity() {
    private lateinit var sharedPrefManager: SharedPrefManager
    private lateinit var mcon: Context
    private val allData = mutableListOf<ChannelsData>()
    private var searchData: List<ChannelsData>? = null
    private var all = true
    private lateinit var adapter: ChannelsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_channels)
        supportActionBar?.elevation = 0f
        supportActionBar?.setBackgroundDrawable(Color.TRANSPARENT.toDrawable())

        val b = intent.extras
        supportActionBar?.subtitle = b?.getString("title")
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        mcon = this

        val decorView = window.decorView
        val wic = WindowInsetsControllerCompat(window, decorView)
        wic.isAppearanceLightStatusBars = false
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            decorView.isForceDarkAllowed = true
        }

        sharedPrefManager = SharedPrefManager(this)
        adapter = ChannelsAdapter(this)

        val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        val rv = findViewById<RecyclerView>(R.id.rv)
        rv.layoutManager = linearLayoutManager
        rv.itemAnimator = DefaultItemAnimator()
        rv.adapter = adapter
        adapter.clear()
        jsonTogson()

        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                finish()
            }
        })
    }

    private fun jsonTogson() {
        val listType = object : TypeToken<List<ChannelsData>>() {}.type
        val gson = Gson()
        val data: List<ChannelsData> = gson.fromJson(sharedPrefManager.getSpChannels(), listType)
        allData.clear()
        allData.addAll(data)
        adapter.addAll(allData)
    }

    private fun filter(text: String) {
        searchData = allData.filter { it.name.lowercase().contains(text.lowercase()) }
        adapter.addAll(searchData!!)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        menuInflater.inflate(R.menu.menu_download, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val searchView = menu.findItem(R.id.action_search).actionView as SearchView
        searchView.queryHint = "Search..."
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean = false

            override fun onQueryTextChange(s: String): Boolean {
                adapter.clear()
                if (s.isEmpty()) {
                    all = true
                    adapter.addAll(allData)
                } else {
                    all = false
                    filter(s)
                }
                return true
            }
        })
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        if (item.itemId == R.id.download) {
            val uri = "https://github.com/bachors/IPTV-Android".toUri()
            startActivity(Intent(Intent.ACTION_VIEW, uri))
        }
        return super.onOptionsItemSelected(item)
    }
}